optaplanner-core/src/main/java/org/optaplanner/core/api/domain/variable/VariableListener.java

```diff
@@ -27,10 +30,14 @@
  * It can change its shadow variable(s) on multiple entity instances
  * (for example: an arrivalTime change affects all trailing entities too).
  * <p>
- * Each {@link ScoreDirector} has a different {@link VariableListener} instance, so it can be stateful.
- * If it is stateful, it must implement {@link StatefulVariableListener}.
+ * It is recommended that implementations be kept stateless.
+ * If state must be implemented, implementations may need to override the default methods
+ * ({@link #resetWorkingSolution(ScoreDirector)}, {@link #close()}).
+ *
+ * @param <Solution_> @{@link PlanningSolution} holding the @{@link PlanningEntity} on which the variable is declared
+ * @param <Entity_> @{@link PlanningEntity} on which the variable is declared
  */
-public interface VariableListener<Solution_, Entity_> extends Supply {
+public interface VariableListener<Solution_, Entity_> extends Closeable {
```

Comment:
Not `AutoClosable`, as I don't intend this API to be used in `try (...)`. It makes no sense.